utils.py

```diff
@@ -0,0 +1,29 @@
+
+class TrainState:
+    """Track number of steps, examples, and tokens processed"""
+
+    step: int = 0  # Steps in the current epoch
+    accum_step: int = 0  # Number of gradient accumulation steps
+    samples: int = 0  # total # of examples used
+    tokens: int = 0  # total # of tokens processed
+
+
+class SimpleLossCompute:
+    """A simple loss compute and train function."""
+
+    def __init__(self, generator, criterion):
+        self.generator = generator
+        self.criterion = criterion
+
+    def __call__(self, x, y, norm):
+        x = self.generator(x)
+        sloss = self.criterion(x.contiguous().view(-1, x.size(-1)), y.contiguous().view(-1)) / norm
+        return sloss.data * norm, sloss
```

Comment:
```suggestion
        return loss
```

Not sure if I'm reading this right, but looks like your going through a lot of trouble here to track unnormalized loss, that you later accumulate and normalize anyway :)